### DISCUSSION TITLE
Should the docstring for a pointers indicate that None is allowed?

### DISCUSSION THREAD

**Discussion by watsonjj** [pybind/pybind11#5059] [Ideas]
Similar to https://github.com/sizmailov/pybind11-stubgen/issues/181, I am trying to indicate in the type hints of my pybind11-wrapped code that a value of None is permitted, as this will be converted to nullptr by pybind11. 

Currently the pybind11 docstring of a pointer or smart pointer is simply the contained type. 

To bypass the current default `name` for a `shared_ptr<Foo>`, I can use the following:

```C++
// Redefine the docstring signature to account for None (nullptr) input
namespace pybind11 { namespace detail {
template<>
struct type_caster<std::shared_ptr<Foo>> : public copyable_holder_caster<Foo, std::shared_ptr<Foo>> {
public:
    using base = type_caster_base<Foo>;
    static constexpr auto name = const_name("Optional[") + base::name + const_name("]");
};
}}
```

1. Is there a trick I am missing here? I can't find a simpler way to override the generated docstring
2. Would a PR which changes the `name` for `type_caster<std::shared_ptr<T>>` and `type_caster<std::unique_ptr<T>>` in cast.h be welcome?
3. I am not sure where the `name` can be overridden for raw pointers...

**Comment by VelocityRa** [comment]
Agreed, people have been requesting this for years (eg. https://github.com/pybind/pybind11/issues/3175, https://github.com/pybind/pybind11/discussions/3377).

Pointers in C/C++ can be null. Seems rather a basic aspect to support.
In C++ people usually rely on that and even intentionally choose pointers over references for that purpose.
Why don't pybind types reflect this, even as an option?
